# DANIEL ONIYIDE Quiz Report

In [28]:
import pandas as pd
import numpy as np

In [2]:
#Question 1: If you're given two lists: A = [1,2,3,4,5,6], B = [13,21,34]. The tasks is to create a list with the elements of A and B in a single dimension with output: A_B = [1,2,3,4,5,6,13,21,34]. Which of the following option is the best way too create this list?

A = [1,2,3,4,5,6]
B = [13,21,34]
print(A)
print(B)


[1, 2, 3, 4, 5, 6]
[13, 21, 34]


Answer: Out of all the options, A.append(B) only updated the A with the parenthesis while A.extend(B) only extended A.
A_B can only be achieve when A and B are added i.e A_B = A + B

In [3]:
A_B = A + B
print(A_B)

[1, 2, 3, 4, 5, 6, 13, 21, 34]


In [4]:
#Question 2: How do you create an identity matrix in python?

C  = np.identity(3)
print(C)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


Answer: Identity matrix can be created using either np.identity(3) or np.eye(3)
From other options, they dont represent well the dimension

In [9]:
Fuel_Data = pd.read_csv(r'C:\Users\Dnear\Desktop\HAMOYE\GROUP A - Introduction to Python for Machine Learning\Lesson 2\fuel_ferc1.csv')

In [16]:
Fuel_Data.rename(columns={'record_id': 'Record ID', 'utility_id_ferc1':'Utility ID','report_year':'Report Year','plant_name_ferc1':'Plant Name','fuel_type_code_pudl':'Fuel Type','fuel_unit':'Fuel Unit','fuel_qty_burned':'Fuel Quantity Burned','fuel_mmbtu_per_unit':'Energy Per Unit','fuel_cost_per_unit_burned':'Fuel Cost Per Unit Burned','fuel_cost_per_unit_delivered':'Fuel Delivery Per Unit','fuel_cost_per_mmbtu':'Fuel Cost Per MMBTU'},inplace=True)
Fuel_Data['Report Year']=pd.to_datetime(Fuel_Data['Report Year'])
Fuel_Data.head()

Utility ID                   Report Year            Plant Name Fuel Type  \
0           1 1970-01-01 00:00:00.000001994              rockport      coal   
1           1 1970-01-01 00:00:00.000001994  rockport total plant      coal   
2           2 1970-01-01 00:00:00.000001994                gorgas      coal   
3           2 1970-01-01 00:00:00.000001994                 barry      coal   
4           2 1970-01-01 00:00:00.000001994             chickasaw       gas   

  Fuel Unit  Fuel Quantity Burned  Energy Per Unit  Fuel Cost Per Unit Burned  \
0       ton             5377489.0           16.590                      18.59   
1       ton            10486945.0           16.592                      18.58   
2       ton             2978683.0           24.130                      39.72   
3       ton             3739484.0           23.950                      47.21   
4       mcf               40533.0            1.000                       2.77   

   Fuel Delivery Per Unit  Fuel Cost Per MMBTU  
0                   18.53                1.121  
1                   18.53                1.120  
2                   38.12                1.650  
3                   45.99                1.970  
4                    2.77                2.570

In [17]:
#Question 3: Which of the following fuel type code has the lowest average fuel cost per unit burned?

Fuel_Data.groupby('Fuel Type')['Fuel Cost Per Unit Burned'].mean()

Fuel Type
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: Fuel Cost Per Unit Burned, dtype: float64

Answer: Gas has the least fuel cost per unit burned of 13.659397.

In [18]:
#Question 4: What is the standard deviation and 75th percentile of the measure of energy per unit in two decimal places?

Fuel_Data['Energy Per Unit'].describe()

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: Energy Per Unit, dtype: float64

Answer: Standard Deviation: 10.60 and 75% percentile:17.01

In [19]:
#Question 5: Wjat is the skewness and kurtosis for the fuel quantity burned in two decimal places?

print(Fuel_Data['Fuel Quantity Burned'].skew())
print(Fuel_Data['Fuel Quantity Burned'].kurt())

15.851495469109503
651.3694501337732


Answer: Skewness: 15.85, Kurtosis: 651.37

In [20]:
#Question 6: What feature has missing values and wha is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places?

Fuel_Data.isna().sum()

Utility ID                     0
Report Year                    0
Plant Name                     0
Fuel Type                      0
Fuel Unit                    180
Fuel Quantity Burned           0
Energy Per Unit                0
Fuel Cost Per Unit Burned      0
Fuel Delivery Per Unit         0
Fuel Cost Per MMBTU            0
dtype: int64

Answer: Feature: Fuel Unit, 
        Total: 180, 
        Percent: 0.610

In [21]:
#Question 7: The feature with missing values falls under what category? What missing value imputation technique would you use?

Fuel_Data.isnull().sum()

Utility ID                     0
Report Year                    0
Plant Name                     0
Fuel Type                      0
Fuel Unit                    180
Fuel Quantity Burned           0
Energy Per Unit                0
Fuel Cost Per Unit Burned      0
Fuel Delivery Per Unit         0
Fuel Cost Per MMBTU            0
dtype: int64

Answer: Categorical and mode imputation

In [22]:
#Question 8: Which of the features has the second and third lowest correlation with the fuel cost per unit burned?

Fuel_Data.corr()

Utility ID  Fuel Quantity Burned  Energy Per Unit  \
Utility ID                   1.000000             -0.057447        -0.066946   
Fuel Quantity Burned        -0.057447              1.000000        -0.080946   
Energy Per Unit             -0.066946             -0.080946         1.000000   
Fuel Cost Per Unit Burned   -0.037863             -0.018535        -0.010034   
Fuel Delivery Per Unit      -0.016414             -0.003551        -0.009039   
Fuel Cost Per MMBTU          0.006122             -0.001896        -0.005884   

                           Fuel Cost Per Unit Burned  Fuel Delivery Per Unit  \
Utility ID                                 -0.037863               -0.016414   
Fuel Quantity Burned                       -0.018535               -0.003551   
Energy Per Unit                            -0.010034               -0.009039   
Fuel Cost Per Unit Burned                   1.000000                0.011007   
Fuel Delivery Per Unit                      0.011007                1.000000   
Fuel Cost Per MMBTU                        -0.000437               -0.000109   

                           Fuel Cost Per MMBTU  
Utility ID                            0.006122  
Fuel Quantity Burned                 -0.001896  
Energy Per Unit                      -0.005884  
Fuel Cost Per Unit Burned            -0.000437  
Fuel Delivery Per Unit               -0.000109  
Fuel Cost Per MMBTU                   1.000000

Answer: The correlation between Fuel cost Per Unit Burned with the other series shows that Fuel Quantity burned (-0.019) has the second lowest value while Fuel Energy per unit (-0.01) has the third lowest value

In [24]:
#Question 9: For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

Fuel_Data.groupby(['Fuel Type','Report Year'])['Fuel Type','Fuel Cost Per Unit Burned'].var()

C:\Users\Dnear\anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  This is separate from the ipykernel package so we can avoid doing imports until


Fuel Cost Per Unit Burned
Fuel Type Report Year                                             
coal      1970-01-01 00:00:00.000001994               1.136227e+02
          1970-01-01 00:00:00.000001995               1.160954e+02
          1970-01-01 00:00:00.000001996               1.055537e+02
          1970-01-01 00:00:00.000001997               2.426727e+06
          1970-01-01 00:00:00.000001998               1.020030e+02
...                                                            ...
waste     1970-01-01 00:00:00.000002014               1.201198e+02
          1970-01-01 00:00:00.000002015               2.756166e+02
          1970-01-01 00:00:00.000002016               3.082602e+02
          1970-01-01 00:00:00.000002017               3.190810e+02
          1970-01-01 00:00:00.000002018               1.188148e+03

[148 rows x 1 columns]

In [25]:
#Question 10: Which year has the highest average fuel cost per unit delivered?

Fuel_Data.groupby('Report Year')['Fuel Delivery Per Unit'].mean()

Report Year
1970-01-01 00:00:00.000001994       63.636060
1970-01-01 00:00:00.000001995       32.735269
1970-01-01 00:00:00.000001996     9196.705948
1970-01-01 00:00:00.000001997    11140.197239
1970-01-01 00:00:00.000001998      287.154420
1970-01-01 00:00:00.000001999       25.551627
1970-01-01 00:00:00.000002000      985.362877
1970-01-01 00:00:00.000002001       60.050396
1970-01-01 00:00:00.000002002       47.594361
1970-01-01 00:00:00.000002003       55.663493
1970-01-01 00:00:00.000002004      139.524275
1970-01-01 00:00:00.000002005       41.438184
1970-01-01 00:00:00.000002006       38.657484
1970-01-01 00:00:00.000002007       43.325023
1970-01-01 00:00:00.000002008       58.588197
1970-01-01 00:00:00.000002009      652.694163
1970-01-01 00:00:00.000002010       91.862105
1970-01-01 00:00:00.000002011       59.774667
1970-01-01 00:00:00.000002012       60.994502
1970-01-01 00:00:00.000002013      172.307591
1970-01-01 00:00:00.000002014      192.737183
1970-01-01 00:00:00.00

Answer: Year 1997 has the higest average fuel cost per unit delivered of 11140.197 units